// Copyright 2022 Google LLC. All Rights Reserved.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
syntax = "proto3";
// All generated protos should be opaque, with "xxx_hidden_" prepended to their field names.


package dcl;

import "proto/connector/sdk.proto";
import "proto/empty.proto";


enum VertexModelSupportedExportFormatsExportableContentsEnum {
  VertexModelSupportedExportFormatsExportableContentsEnumNO_VALUE_DO_NOT_USE = 0;
  VertexModelSupportedExportFormatsExportableContentsEnumEXPORTABLE_CONTENT_UNSPECIFIED = 1;
  VertexModelSupportedExportFormatsExportableContentsEnumARTIFACT = 2;
  VertexModelSupportedExportFormatsExportableContentsEnumIMAGE = 3;
}

enum VertexModelSupportedDeploymentResourcesTypesEnum {
  VertexModelSupportedDeploymentResourcesTypesEnumNO_VALUE_DO_NOT_USE = 0;
  VertexModelSupportedDeploymentResourcesTypesEnumDEPLOYMENT_RESOURCES_TYPE_UNSPECIFIED = 1;
  VertexModelSupportedDeploymentResourcesTypesEnumDEDICATED_RESOURCES = 2;
  VertexModelSupportedDeploymentResourcesTypesEnumAUTOMATIC_RESOURCES = 3;
}

message VertexModel {
  string name = 1;
  string version_id = 2;
  string version_create_time = 3;
  string version_update_time = 4;
  string display_name = 5;
  string description = 6;
  string version_description = 7;
  repeated VertexModelSupportedExportFormats supported_export_formats = 8;
  string training_pipeline = 9;
  VertexModelOriginalModelInfo original_model_info = 10;
  VertexModelContainerSpec container_spec = 11;
  string artifact_uri = 12;
  repeated VertexModelSupportedDeploymentResourcesTypesEnum supported_deployment_resources_types = 13;
  repeated string supported_input_storage_formats = 14;
  repeated string supported_output_storage_formats = 15;
  string create_time = 16;
  string update_time = 17;
  repeated VertexModelDeployedModels deployed_models = 18;
  string etag = 19;
  map<string, string> labels = 20;
  VertexModelEncryptionSpec encryption_spec = 21;
  string project = 22;
  string location = 23;
}

message VertexModelSupportedExportFormats {
  string id = 1;
  repeated VertexModelSupportedExportFormatsExportableContentsEnum exportable_contents = 2;
}

message VertexModelOriginalModelInfo {
  string model = 1;
}

message VertexModelContainerSpec {
  string image_uri = 1;
  repeated string command = 2;
  repeated string args = 3;
  repeated VertexModelContainerSpecEnv env = 4;
  repeated VertexModelContainerSpecPorts ports = 5;
  string predict_route = 6;
  string health_route = 7;
}

message VertexModelContainerSpecEnv {
  string name = 1;
  string value = 2;
}

message VertexModelContainerSpecPorts {
  int64 container_port = 1;
}

message VertexModelDeployedModels {
  string endpoint = 1;
  string deployed_model_id = 2;
}

message VertexModelEncryptionSpec {
  string kms_key_name = 1;
}

message ApplyVertexModelRequest {
  VertexModel resource = 1;
  repeated LifecycleDirective lifecycle_directives = 2;
  string service_account_file = 3;
}

message DeleteVertexModelRequest {
  string service_account_file = 1;
  VertexModel resource = 2;
}

message ListVertexModelRequest {
  string service_account_file = 1;
  string Project = 2;
  string Location = 3;
}

message ListVertexModelResponse {
  repeated VertexModel items = 1;
}

service VertexModelService {
  rpc ApplyVertexModel(ApplyVertexModelRequest) returns (VertexModel);
  rpc DeleteVertexModel(DeleteVertexModelRequest) returns (google.protobuf.Empty);
  rpc ListVertexModel(ListVertexModelRequest) returns (ListVertexModelResponse);
}
